File: src/main/java/TeamManager.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class TeamManager extends AbstractEmployee implements Manager{
    private final List<Employee> employeeList = new ArrayList<>();
    private final Integer subordinateNumber;
    private final Random gen = new Random();

    public TeamManager(String name, String role, int subordinateNumber)
    {
        super(name, role);
        if (subordinateNumber > 0) this.subordinateNumber = subordinateNumber;
        else throw new IllegalArgumentException("Invalid number of subordinates: " + subordinateNumber);
    }

    @Override
    public boolean canHire() {
        return subordinateNumber > employeeList.size();
    }

    @Override
    public void fire(Employee employee) {
        employeeList.remove(employee);
        employee.setSuperior(null);
    }

    @Override
    public void hire(Employee employee) {
        employeeList.add(employee);
        employee.setSuperior(this);
    }

    public void assaign(Task task)
    {
        if (!employeeList.isEmpty()) employeeList.get(gen.nextInt(employeeList.size())).assaign(task);
        else System.out.println("Cannot assign task " + task.getTaskName());
    }

    public Report reportWork()
    {
        Report report = new Report(new LinkedList<>());
        for (Employee employee : employeeList) report = report.concatenate(employee.reportWork());
        return report;
    }
}

File: src/main/scala/data/domain/TeamManager.scala
package data.domain

import data.domain.Gender.Gender

import scala.util.Random

class TeamManager(firstName: String,
                  lastName: String,
                  role: String,
                  email: Email,
                  gender: Gender,
                  country: String,
                  university: String,
                  supervisor: Manager,
                  company: Company,
                  var hiringLimit: Int)
  extends AbstractEmployee(firstName, lastName, role, email, gender, country, university, supervisor, company) with Manager {

  private val rnd = new Random

  var employees : Set[Employee] = Set()

  override var hiringPredicate: Employee => Boolean = emp => true

  override def canHire: Boolean = employees.size < hiringLimit

  override def fire(employee: Employee): Unit = {
    if (employees.contains(employee)) employees = employees - employee
  }

  override def hire(employee: Employee): Unit = {
    if (canHire && hiringPredicate(employee)) {
      employees = employees + employee
      employee.supervisor = this
      employee.company = company
    }
  }

  override def reportWork: Report = employees.toList.foldLeft(new Report(List()))(_ + _.reportWork)

  override def assign(task: Task): Unit = employees.toVector(rnd.nextInt(employees.size)).assign(task)

  override def totalNumberOfEmployees: Int = employees.size + employees.filter(_.isInstanceOf[Manager]).foldLeft(0)(_ + _.asInstanceOf[Manager].totalNumberOfEmployees)

  override def allEmployees: Set[Employee] = employees ++ employees.filter(_.isInstanceOf[Manager]).foldLeft(Set[Employee]())(_ ++ _.asInstanceOf[Manager].allEmployees)

  override def allManagers: Set[Manager] = Set(this) ++ employees.filter(_.isInstanceOf[Manager]).foldLeft(Set[Manager]())(_ ++ _.asInstanceOf[Manager].allManagers)

  override def clone(employee: Employee): Unit = {
    super.clone(employee)
    hiringPredicate = employee.asInstanceOf[Manager].hiringPredicate
    hiringLimit = employee.asInstanceOf[Manager].hiringLimit
    employees = employee.asInstanceOf[Manager].employees
  }
}
